### PR TITLE
doc(ci): document GCB testing principles

### PR DESCRIPTION
**Pull request by devjgm**
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6131)

### CODE REVIEW

File: ci/cloudbuild/README.md [hunk 1]
```diff
@@ -1,23 +1,51 @@
...
+
+GCB triggers can be configured in the http://console.cloud.google.com/ UI, but
+we prefer to configure them with SCM controlled YAML files that live in the
+`triggers/` directory. Managing triggers can be done with the `gcloud` command
+also, but we have the local `trigger.sh` script to make this process a bit
+easier. See `trigger.sh --help` for more details.
+
+## Testing principles
+
+We want our code to work for our customers. We don't their exact environment
```
Comment (coryan) [inline ci/cloudbuild/README.md:19]: s/don't/don't know/ ?

File: ci/cloudbuild/README.md [hunk 2]
```diff
@@ -1,23 +1,51 @@
...
+and configuration, so we need to test our code in a variety of different
+environments and configurations. The main dimensions that we need to test are:
+
+- OS Platform: Windows, macOS, Linux x `N` different distros
+- Compilers: clang, gcc, msvc
```
Comment (coryan) [inline ci/cloudbuild/README.md:24]: I was told those are supposed to be spelled Clang, GCC and MSVC.

File: ci/cloudbuild/README.md [hunk 3]
```diff
@@ -1,23 +1,51 @@
...
+- Build tool: bazel, CMake
+- Configuration: release, debug, with/without exceptions, dynamic vs static
+- C++ Language: C++11, ..., C++17 (soon to be C++20)
```
Comment (coryan) [inline ci/cloudbuild/README.md:27]: s/soon to be C++20/soon we will need to add C++20/ ?

File: ci/cloudbuild/README.md [hunk 4]
```diff
@@ -1,23 +1,51 @@
...
+- Installation: cmake, pkg-config
+
+In addition to these main dimensions, we also want to use tools and analyzers
+to help us catch bugs: clang-tidy, clang static analyzer, asan, msan, tsan,
+ubsan, etc. The full matrix of all combinations is infeasible to test
+completely, so we follow the following principles to minimize the test space
+while achieving high likelihood of the code working for our customers.
+
+* We want to test _something_ on each dimension
```
Comment (coryan) [inline ci/cloudbuild/README.md:36]: I think I would say:

* For simple dimensions (e.g. things that are "on/off") we want at least one build for each 'value' of the setting.

**Review by coryan — APPROVED** [review]
A few suggestions below.  I think strictly optional.

File: ci/cloudbuild/README.md (hunk 1 above)
Comment (devjgm) [reply ci/cloudbuild/README.md]: Fixed.

File: ci/cloudbuild/README.md (hunk 2 above)
Comment (devjgm) [reply ci/cloudbuild/README.md]: Fixed there and elsewhere.

File: ci/cloudbuild/README.md (hunk 3 above)
Comment (devjgm) [reply ci/cloudbuild/README.md]: The intent here was to list the range, not the values we test.... and soon the upper end will be C++20... actually that's true now, I'll just change it.

File: ci/cloudbuild/README.md (hunk 4 above)
Comment (devjgm) [reply ci/cloudbuild/README.md]: Almost missed this. Done.

**Pull request merged**